### ISSUE TITLE
[train] Incorrect advantage assignment for step_wise_trajectores in RayPPOTrainer?

### ISSUE THREAD

**Issue by rehaanahmad2013** [NovaSky-AI/SkyRL#1492]
I've been having some stability issues when training with step_wise_trajectories. I see in the RayPPOTrainer we are doing 

```
traj_ids = (
    torch.cat([torch.tensor([False], device=is_last_step.device), is_last_step[:-1]]).int().cumsum(dim=0)
)
num_groups = traj_ids[-1].item() + 1
assert num_groups == len(
    last_step_advantages
), f"number of groups {num_groups} doesn't match the number of trajectories as given by `is_last_step` {len(last_step_advantages)}. The `is_last_step` tensor is likely malformed"
advantages = last_step_advantages[traj_ids]
returns = last_step_returns[traj_ids]
```

[link](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/train/trainer.py#L823)

But assigning `advantages = last_step_advantages[traj_ids]` doesn't seem quite right to me since `last_step_advantages` is a `scalar * response_mask` of the last step. This means the previous steps in a trajectory are getting an advantage with a different response mask. This is especially problematic if the last step `response_ids` is short and the prior steps see much longer obs_ids (thus there is virtually no overlap in action tokens due to the left padding). Shouldn't we be getting the advantage of the last step as a scalar and multiplying that by `data["response_mask"]`? 

When making this fix I'm seeing much better stability, but do let me know if I'm looking at this correctly. Also happy to open a small PR if that helps.

**Comment by CharlieFRuan** [comment]
@rehaanahmad2013 Thank you so much for raising it, should be fixed via: https://github.com/NovaSky-AI/SkyRL/pull/1507

Let us know if this fix gets you the same stability with your local fix when you get a chance. Much appreciated!

Note that at HEAD, step-wise might run into this issue https://github.com/NovaSky-AI/SkyRL/issues/1501, which will be gone after merging https://github.com/NovaSky-AI/SkyRL/pull/1483, or a quick padding hot fix described in https://github.com/NovaSky-AI/SkyRL/pull/1507

Apologies for the friction! The step-wise training codepath is actively being polished: https://github.com/NovaSky-AI/SkyRL/issues/1278

**Comment by rehaanahmad2013** [comment]
@CharlieFRuan  Thank you! Will keep you posted. 

Btw do you typically run with advantage_batch_normalize? Have you found that to be useful?

**Comment by SumanthRH** [comment]
> Btw do you typically run with advantage_batch_normalize? Have you found that to be useful?

We typically run without advantage_batch_normalize. I experimented with advantage_batch_normalize long back for single turn training and didn't find it useful then. I think we just kept it as an option because it was also an option in classical RLHF libraries.

**Comment by SumanthRH** [comment]
Have you experimented with it for your step wise training runs @rehaanahmad2013 ?

**Comment by rehaanahmad2013** [comment]
Running rn @SumanthRH ! Will keep you posted

**Comment by rehaanahmad2013** [comment]
Hmm I'm still running into some instability/collapse when setting stepwise to true. Have you needed to tune LR/kl coef when doing these stepwise runs @SumanthRH ?

**Comment by CharlieFRuan** [comment]
Hi @rehaanahmad2013 ! Here are some curves where we use step-wise for search-r1: https://wandb.ai/sky-posttraining-uc-berkeley/skyrl-search-padding/reports/PR-1285-search-r1-PAD-left-only-baseline-after-PR-and-step-wise---VmlldzoxNjE3MTA0OQ

The run `stepwise-pr1507-pad_batchHotFix` is run right after this PR is merged and with the description here: https://github.com/NovaSky-AI/SkyRL/pull/1507

The configs are exactly the same as https://github.com/NovaSky-AI/SkyRL/blob/main/examples/train/search/run_search.sh but with step-wise ON.

Does your training collapse with stepwise, but stable without stepwise?

**Comment by rehaanahmad2013** [comment]
Thanks for sharing (I can't access the wandb link but I believe you haha). Yes my training is stable without stepwise but pretty inconsistent with stepwise.

I've been running with #1483 (+ my local fix) perhaps I should just properly merge from main with the #1507 change as you suggest.

**Comment by CharlieFRuan** [comment]
Oh shoot, here is the link: https://api.wandb.ai/links/sky-posttraining-uc-berkeley/h564n3ln

#1483 will be merged tomorrow for sure, after that you can run at main. Apologies for the delay!

**Comment by CharlieFRuan** [comment]
@rehaanahmad2013 Merged! https://github.com/NovaSky-AI/SkyRL/pull/1529

Step-wise training at HEAD will work now. You can see the commands I used to produce the search-r1 curves in the PR description.

**Comment by rehaanahmad2013** [comment]
@CharlieFRuan I'm getting some stable results with stepwise now, thanks!

**Comment by CharlieFRuan** [comment]
@rehaanahmad2013 Nice! So glad to hear!
